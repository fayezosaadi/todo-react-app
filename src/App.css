.App {
  text-align: center;
  background-color: white;
  margin: 0 20%;
}

.todoListMain .todoListHeader {
  display: flex;
  justify-content: space-between;
  padding: 0 12px;
  background-color: aliceblue;
  height: 68px;
}
.todoListMain .todoListHeader form {
  margin-block-start: 1em;
  margin-block-end: 1em;
  margin-inline-start: 0;
  margin-inline-end: 0;
}
.todoListMain .todoListHeader form label {
  font-size: small;
}

.todoListMain .todoInputForm {
  display: flex;
  justify-content: space-between;
}
.todoListMain .todoInputForm input {
  padding: 10px;
  font-size: 16px;
  border: 2px solid #FFF;
  outline: none;
  width: 100%;
}

.todoListMain .theList {
  padding-inline-start: 0;
  width: 100%;
}
.todoListMain .theList .listItem {
  display: flex;
  justify-content: space-between;
}
.todoListMain .theList form {
  border-bottom: 1px solid darkgrey;
  margin: 10px 25px
}
.todoListMain .theList .listItem button {
  border: none;
  font-weight: bold;
  outline: none;
}
.todoListMain .theList .listItem .listItemText {
  width: 90%;
  display: flex;
}
.todoListMain .theList .listItem .listItemText input {
  margin-right: 13px;
}
