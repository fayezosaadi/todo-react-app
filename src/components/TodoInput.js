import React from 'react';

const TodoInput = ({ addItem, handleInput, text }) => {
  return (
      <div>
        <form className="todoInputForm" onSubmit={addItem}>
          <input placeholder="Type to add new tasks" onChange={handleInput} value={text} />
        </form>
    </div>
  );
};

export default TodoInput;
