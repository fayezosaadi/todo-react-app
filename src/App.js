import React, { useState, useEffect } from 'react';
import { v4 as uuid } from 'uuid';
import './App.css';
import TodoInput from "./components/TodoInput";
import TodoItems from "./components/TodoItems";
import TodoHeader from "./components/TodoHeader";
const localStorage = require("local-storage");

const currentItemInitialState = { text: '', id: '', isCompleted: false };

const useStateWithLocalStorage = localStorageKey => {
  const [items, setItems] = useState(JSON.parse(localStorage.get(localStorageKey)) || []);

  useEffect(() => { localStorage.set('items', JSON.stringify(items)) }, [items]);

  return [items, setItems];
};

const App = () => {
  const [ items, setItems ] = useStateWithLocalStorage('items');
  const [ hide, setHide ] = useState(false );
  const [ currentItem, setCurrentItem ] = useState(currentItemInitialState);

  const handleInput = e => {
    const { target: { value } } = e;
    setCurrentItem(item => ({ ...item, text: value, id: uuid() }));
};

  const addItem = e => {
    e.preventDefault();

    if (currentItem.text) {
      setItems(itemsList => [...itemsList, currentItem]);
      setCurrentItem(currentItemInitialState);
    }
  };

  const deleteItem = (e, id) => {
    e.preventDefault();

    setItems(items.filter(item => item.id !== id))
  };

  const completeTodo = (e, i) => {
    const { target: { type, checked, value, name } } = e;
    const newTodos = [...items];
    newTodos[i][name] = type === 'checkbox' ? checked : value;
    setItems(newTodos)
  };

  const hideCompleted = (e) => {
    const { target: { type, checked, value } } = e;
    setHide(type === 'checkbox' ? checked : value)
  };

  return (
    <div className="App">
      <div className="todoListMain">
        <TodoHeader hide={hide} hideCompleted={hideCompleted} />
        <TodoInput addItem={addItem} handleInput={handleInput} currentItem={currentItem} />
        <TodoItems items={items} deleteItem={deleteItem} completeTodo={completeTodo} hide={hide} />
      </div>
    </div>
  );
};

export default App;
