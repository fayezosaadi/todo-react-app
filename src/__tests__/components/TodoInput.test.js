import React from "react";
import { fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import TodoInput from "../../components/TodoInput";
import App from "../../App";

const localStorage = require("local-storage");

describe('<TodoInput />', () => {
  it('should render', () => {
    render(<TodoInput />)
  });

  it('should render and match the snapshot', () => {
    const component = renderer.create(<TodoInput />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should update the state', () => {
    const { getByPlaceholderText } = render(<App />);
    expect(getByPlaceholderText(/Type to add new tasks/i).textContent).toBe("");
    fireEvent.change(getByPlaceholderText("Type to add new tasks"), { target: { value: 'First Todo' } } )
    expect(getByPlaceholderText(/Type to add new tasks/i).value).toBe("First Todo");
  });

  it("should submit and update state value", () => {
    const { getByPlaceholderText, getByText } = render(<App />);
    const input = getByPlaceholderText(/Type to add new tasks/i);
    fireEvent.change(input, { target: { value: 'First Todo' } } );
    fireEvent.submit(input);
    const text = getByText(/First Todo/i);
    expect(text).toBeInTheDocument();
    const [item] = JSON.parse(localStorage.get("items"));
    expect(item.text).toBe("First Todo");
  })

});
