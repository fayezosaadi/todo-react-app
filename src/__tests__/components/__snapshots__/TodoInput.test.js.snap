// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoInput /> should render and match the snapshot 1`] = `
<div>
  <form
    className="todoInputForm"
  >
    <input
      placeholder="Type to add new tasks"
    />
  </form>
</div>
`;
